### ISSUE TITLE
[Python] Unable to deserialize when RestAPI answer is a subtype

### ISSUE THREAD

**Issue by lmazuel** [Azure/autorest#1729]
In [Machine Learning Swagger](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-machinelearning/2016-05-01-preview/swagger/webservices.json): 

```json
    "WebServiceProperties": {
      "type": "object",
      "description": "The set of properties specific to the Azure ML web service resource.",
      "required": [
        "packageType"
      ],
      "discriminator": "packageType",
.......
        "packageType": {
          "type": "string",
          "description": "Specifies the package type. Valid values are Graph (Specifies a web service published through the Machine Learning Studio) and Code (Specifies a web service published using code such as Python). Note: Code is not supported at this time.",
          "enum": [
            "Graph"
          ]        
        }
      }
    },
    "WebServicePropertiesForGraph": {
      "description": "Properties specific to a Graph based web service.",
      "x-ms-discriminator-value": "Graph",
      "allOf": [
        {
          "$ref": "#/definitions/WebServiceProperties"
        }
      ],
```

when deserializing an answer from the RestAPI, it fails:
```python
Traceback (most recent call last):
  File "D:\VSProjects\msrest-for-python\msrest\serialization.py", line 837, in _instantiate_model
    response_obj = response(**kwargs)
TypeError: __init__() got an unexpected keyword argument 'package_type'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "d:\vsprojects\azure-sdk-for-python\azure-mgmt\tests\test_mgmt_machinelearning.py", line 288, in test_ml
    webservice = async_webservice.result()
  File "D:\VSProjects\msrestazure-for-python\msrestazure\azure_operation.py", line 473, in result
    self.wait(timeout)
  File "D:\VSProjects\msrestazure-for-python\msrestazure\azure_operation.py", line 486, in wait
    raise self._exception
  File "D:\VSProjects\msrestazure-for-python\msrestazure\azure_operation.py", line 367, in _start
    self._poll(update_cmd)
  File "D:\VSProjects\msrestazure-for-python\msrestazure\azure_operation.py", line 463, in _poll
    self._response)
  File "D:\VSProjects\msrestazure-for-python\msrestazure\azure_operation.py", line 292, in get_status_from_resource
    self.resource = self._deserialize(response)
  File "D:\VSProjects\msrestazure-for-python\msrestazure\azure_operation.py", line 177, in _deserialize
    resource = self.get_outputs(response)
  File "d:\vsprojects\azure-sdk-for-python\azure-mgmt-machinelearning\azure\mgmt\machinelearning\webservices\operations\web_services_operations.py", line 117, in get_long_running_output
    deserialized = self._deserialize('WebService', response)
  File "D:\VSProjects\msrest-for-python\msrest\serialization.py", line 762, in __call__
    value = self.deserialize_data(raw_value, attr_type)
  File "D:\VSProjects\msrest-for-python\msrest\serialization.py", line 888, in deserialize_data
    return self(obj_type, data)
  File "D:\VSProjects\msrest-for-python\msrest\serialization.py", line 768, in __call__
    return self._instantiate_model(response, d_attrs)
  File "D:\VSProjects\msrest-for-python\msrest\serialization.py", line 844, in _instantiate_model
    raise DeserializationError(msg + str(err))
msrest.exceptions.DeserializationError: Unable to deserialize {'realtime_configuration': <azure.mgmt.machinelearning.webservices.models.realtime_configuration.RealtimeConfiguration object at 0x04B14590>, 'description': None, 'output': <azure.mgmt.machinelearning.webservices.models.service_input_output_specification.ServiceInputOutputSpecification object at 0x04B14530>, 'storage_account': <azure.mgmt.machinelearning.webservices.models.storage_account.StorageAccount object at 0x04B14550>, 'read_only': False, 'diagnostics': <azure.mgmt.machinelearning.webservices.models.diagnostics_configuration.DiagnosticsConfiguration object at 0x04B145B0>, 'parameters': {'trigonometric Function': 'Sin', 'output mode': 'ResultOnly', 'category': 'Trigonometric', 'column Set': '{"isFilter":true,"rules":[{"ruleType":"ColumnTypes","exclude":false,"columnTypes":["Numeric"],"columnKinds":["All"]}]}'}, 'commitment_plan': None, 'machine_learning_workspace': None, 'assets': {'asset1': <azure.mgmt.machinelearning.webservices.models.asset_item.AssetItem object at 0x04B14630>}, 'keys': None, 'expose_sample_data': True, 'input': <azure.mgmt.machinelearning.webservices.models.service_input_output_specification.ServiceInputOutputSpecification object at 0x04B14610>, 'title': None, 'example_request': <azure.mgmt.machinelearning.webservices.models.example_request.ExampleRequest object at 0x04B145F0>, 'package_type': 'Graph'} into model <class 'azure.mgmt.machinelearning.webservices.models.web_service_properties.WebServiceProperties'>. __init__() got an unexpected keyword argument 'package_type'
```

[And looking at the code](https://github.com/lmazuel/azure-sdk-for-python/blob/machine_learning_mgmt/azure-mgmt-machinelearning/azure/mgmt/machinelearning/webservices/models/web_service_properties.py), `package_type` is not a parameter of `WebServiceProperties` indeed:
```python
class WebServiceProperties(Model):

    _validation = {
        'created_on': {'readonly': True},
        'modified_on': {'readonly': True},
        'provisioning_state': {'readonly': True},
        'swagger_location': {'readonly': True},
        'package_type': {'required': True},
    }

    _attribute_map = {
...
        'parameters': {'key': 'parameters', 'type': '{str}'},
        'package_type': {'key': 'packageType', 'type': 'str'},
    }

    _subtype_map = {
        'package_type': {'Graph': 'WebServicePropertiesForGraph'}
    }

    def __init__(self, title=None, description=None, keys=None, read_only=None, expose_sample_data=None, realtime_configuration=None, diagnostics=None, storage_account=None, machine_learning_workspace=None, commitment_plan=None, input=None, output=None, example_request=None, assets=None, parameters=None):
...
        self.parameters = parameters
        self.package_type = None
```

**Comment by lmazuel** [comment]
(needs https://github.com/Azure/autorest/pull/1727 to be reproduced)

**Comment by lmazuel** [comment]
@fearthecowboy related to https://github.com/Azure/autorest/issues/1720?

**Comment by lmazuel** [comment]
Got the issue with latest update of Commerce too:
https://github.com/Azure/azure-sdk-for-python/pull/959/files

**Comment by lmazuel** [comment]
@annatisch PR https://github.com/Azure/autorest/pull/1726 didn't solve this, could you check in Python has also an issue?

Thank you,

**Comment by fearthecowboy** [comment]
This isn't the same as #1726...

I have discovered that the different generators handle polymorphism somewhat differently, and it makes me sad.

The  Modeler removes the discriminator parameter from the generated code, but some of the generators look at that and creates a new parameter for the discriminator.

Ideally, this should be something resolved in deserialization, and wouldn't need to actually have the physical parameter in the object (which is what C# does) but different languages, different ideas.... Diversity is a wonderful thing :D

The other fix #1726 was when a model class was empty except for the discriminator (which when it checked to see if it should auto-flatten, zero properties was less than or equal to zero, and so it eliminated the class.

**Comment by lmazuel** [comment]
@fearthecowboy I think I get that you mean.

However I'm not enough "Autorest dev" yet to know what should be the right behavior: does Autorest has a way to know that it should deserialize as the specific class (`WebServicePropertiesForGraph`) and not the generic one (`WebServiceProperties`).

Then, I think we should remove the discrimators parameters from the JSON before deserializing?

**Comment by fearthecowboy** [comment]
Well, that's where it gets kindof funky...

AutoRest doesn't know, the Modeler marks model types with `PolymorphicDiscriminator`, and sets the `SerializedName` to the value that the deserializer in the client uses to determine what type to create a runtime.

In c# this magic is applied with such a light touch, that it was tricky for me to even see how it worked. Specifically, (in c#):
- the modeler marks a composite type with the Polymorphic Discriminator
- the modeler sets the serialized name
- the generator puts an attribute on the generated class with `SerializedName` (and skips that if `SerializedName` == the class name! )
- the generator inserts a couple lines into the client:

```
SerializationSettings.Converters.Add(new Microsoft.Rest.Serialization.PolymorphicSerializeJsonConverter<@(polymorphicType.Name)>("@(polymorphicType.PolymorphicDiscriminator)"));
DeserializationSettings.Converters.Add(new  Microsoft.Rest.Serialization.PolymorphicDeserializeJsonConverter<@(polymorphicType.Name)>("@(polymorphicType.PolymorphicDiscriminator)"));
```

and that's *IT* ... The magic is handled by the client runtime and JSON.NET which is so f'ing subtle that it's creepy.

**Comment by lmazuel** [comment]
Thank you @fearthecowboy that was critical to get it :)
Looking at the [C# code](https://github.com/Azure/azure-sdk-for-net/tree/AutoRest/src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Generated/WebServices) I think I understand what we should do now. JSON deserializer in Python is not smart as JSON.NET, but should be ok :)

**Comment by lmazuel** [comment]
@fearthecowboy last question, is there tests in the test server for that? So I can add them in Python (we clearly miss tests since there is a bug :))

**Comment by annatisch** [comment]
@fearthecowboy, @lmazuel 
As it happens the Python generator is *not* supposed to be generating the polymorphic discriminator as an attribute of the model class - I'm not sure when/how this snuck in (probably very early on) but the runtime is designed to serialize/deserialize without this attribute - so it should be safe to remove it (in fact I have done so, in a long forgotten branch in my autorest fork waiting for me to get back to it).

**Comment by lmazuel** [comment]
@annatisch 
If you look at the traceback,:
```python
msrest.exceptions.DeserializationError: Unable to deserialize {} into model <class 'azure.mgmt.machinelearning.webservices.models.web_service_properties.WebServiceProperties'>. __init__() got an unexpected keyword argument 'package_type'
```

But we should try to deserialize as `WebServicePropertiesForGraph`, not `WebServiceProperties`. So we miss the mechanism described by @fearthecowboy to deduce the subtype.

**Comment by annatisch** [comment]
@lmazuel - yes my post was less about the actual issue - and more in general. In this case a _subtype_map is missing on the parent class.

**Comment by fearthecowboy** [comment]
I don't recall if there are.  Honestly, I'd be surprised if there are, simply because it's kindof  tricky to know if it's done right, (per-language)

**Comment by lmazuel** [comment]
Fixed by msrest 0.4.5
https://pypi.python.org/pypi/msrest/0.4.5
